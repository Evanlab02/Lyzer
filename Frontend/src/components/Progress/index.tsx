import { useMemo } from "react";
import "./styles/index.scss";

export interface ProgressProps {
    value?: number;
    max?: number;
    useProgressColour?: boolean;
    testId?: string;
}

export default function Progress(props: Readonly<ProgressProps>) {
	const { 
		value = 0, 
		max = 100,
		useProgressColour = false,
		testId
	} = props;

	const progressClassName = useMemo(() => {
		const progressPercentage = (value / max) * 100;
		if (useProgressColour && progressPercentage == 100) {
			return "lyzer-progress-complete";
		} else if (useProgressColour && 80 <= progressPercentage && progressPercentage < 100) {
			return "lyzer-progress-almost";
		} else if (useProgressColour && progressPercentage < 80) {
			return "lyzer-progress-incomplete";
		}
	}, [value, max, useProgressColour]);

	return (
		<progress
			className={`lyzer-progress ${progressClassName ?? ""}`}
			value={value}
			max={max}
			data-testid={testId}
		/>
	);
}