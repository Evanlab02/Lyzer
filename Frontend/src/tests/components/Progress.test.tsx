import { expect, it } from "vitest";
import { render } from "@testing-library/react";
import Progress from "../../components/Progress";

it("Empty progress bar renders correctly", async () => {
	const { findByTestId } = render(
		<Progress 
			testId="lyzer-progress-bar"
			value={0}
			max={100}
		/>
	);
	const progressBar = await findByTestId("lyzer-progress-bar");
	expect(progressBar).toMatchSnapshot();
});

it("Progress bar renders correctly with value", async () => {
	const { findByTestId } = render(
		<Progress 
			testId="lyzer-progress-bar"
			value={50}
			max={100}
		/>
	);
	const progressBar = await findByTestId("lyzer-progress-bar");
	expect(progressBar).toMatchSnapshot();
});

it("Progress bar renders correctly with 50% value", async () => {
	const { findByTestId } = render(
		<Progress 
			testId="lyzer-progress-bar"
			value={50}
			max={100}
			useProgressColour
		/>
	);
	const progressBar = await findByTestId("lyzer-progress-bar");
	expect(progressBar).toMatchSnapshot();
});


it("Progress bar renders correctly with 80% value", async () => {
	const { findByTestId } = render(
		<Progress 
			testId="lyzer-progress-bar"
			value={80}
			max={100}
			useProgressColour
		/>
	);
	const progressBar = await findByTestId("lyzer-progress-bar");
	expect(progressBar).toMatchSnapshot();
});

it("Progress bar renders correctly with 100% value", async () => {
	const { findByTestId } = render(
		<Progress 
			testId="lyzer-progress-bar"
			value={100}
			max={100}
			useProgressColour
		/>
	);
	const progressBar = await findByTestId("lyzer-progress-bar");
	expect(progressBar).toMatchSnapshot();
});
