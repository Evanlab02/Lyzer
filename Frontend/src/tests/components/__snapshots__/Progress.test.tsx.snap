// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Empty progress bar renders correctly 1`] = `
<progress
  class="lyzer-progress "
  data-testid="lyzer-progress-bar"
  max="100"
  value="0"
/>
`;

exports[`Progress bar renders correctly with 50% value 1`] = `
<progress
  class="lyzer-progress lyzer-progress-incomplete"
  data-testid="lyzer-progress-bar"
  max="100"
  value="50"
/>
`;

exports[`Progress bar renders correctly with 80% value 1`] = `
<progress
  class="lyzer-progress lyzer-progress-almost"
  data-testid="lyzer-progress-bar"
  max="100"
  value="80"
/>
`;

exports[`Progress bar renders correctly with 100% value 1`] = `
<progress
  class="lyzer-progress lyzer-progress-complete"
  data-testid="lyzer-progress-bar"
  max="100"
  value="100"
/>
`;

exports[`Progress bar renders correctly with value 1`] = `
<progress
  class="lyzer-progress "
  data-testid="lyzer-progress-bar"
  max="100"
  value="50"
/>
`;
